### ISSUE TITLE
Missing unwrap for angle difference

### ISSUE THREAD

**Issue by thomasfermi** [goktug97/PyGraphSLAM#1]
Hi @goktug97,
I think that there might be a problem with this line

https://github.com/goktug97/PyGraphSLAM/blob/33e8a7f33da549a82afdd234d6283a7610b9096c/slam.py#L95

Since `odom = x, y, theta`, your `dx[2]` will bei incorrect whenever theta jumps from -pi to pi or from pi to pi, which definitely happens in the data of Uni Freiburg you are mentioning in the README (see plot below).

A fix would be to put `odoms[:,2] = np.unwrap(odoms[:,2])` right after constructing `odoms`. 

![Screenshot from 2020-03-25 16-46-19](https://user-images.githubusercontent.com/30798048/77556020-39253900-6eb8-11ea-84d7-6b8810977c4c.png)

**Comment by thomasfermi** [comment]
I had written that your `dx[2]` is incorrect. That is not really true. Actually your `dx[2]` is just not properly wrapped.

But I just noticed that the only place where you use `dx[2]` is inside trigonometric function, so it does not matter. sin(x-2pi)=sin(x).

Sorry for raising an issue which is unnecessary.

**Comment by goktug97** [comment]
Thank you for looking into it. I think there are other problems because it can't handle other maps. In my tests, it couldn't map reliably except the intel dataset. Grid mapping might solve the problem

**Comment by thomasfermi** [comment]
Hey @goktug97, thanks for you reply! I am just trying to learn about Graph Slam by reading the [tutorial by Grisetti et al.](http://www2.informatik.uni-freiburg.de/~stachnis/pdf/grisetti10titsmag.pdf) and watching the [lectures by Cyrill Stachniss](https://youtu.be/VRGOLRGwAjg). I found your code to be a very good learning resource, because it is so concise.

I am sorry to hear that it does not work for the other maps. Why do you think that grid mapping might help? Would you also do the icp based on the grid map then?

I'm an absolute beginner, but here are my thoughts on where the code might be improved:
- Have you tried to additionally add edges based on the odometry? These would only be between consecutive vertices of course.
- I see that you are currently using the identity matrix for the covariance. You have some commented code that was trying to estimate the covariance, which is broken according to your comment. Maybe one could try a very simple estimate for the covariance as `c*np.eye(3)` where c is the mean of `distances`. The higher `c` is, the more it seems like the icp did not work so well and we should increase the estimate of the covariance.

EDIT: Just checked the icp code again. The tolerance threshold ensures that the mean of `distances` is always tiny. Maybe one could estimate `c` as `c=1/number_of_matches` and maybe also make sure that `number_of_matches` is not too small. Maybe at least 3.
EDIT2: Maybe one could also increase `c` based on the amount of rotation between scans A and B. If you do ICP along a straight corridor where you move straight, I guess that the result won't be that great. Hence in such situations the covariance should maybe be a bit higher.

**Comment by goktug97** [comment]
Currently, every point is added to the map so there is a lot of noise. Grid mapping would eliminate this noise because the probability of being free will be higher than that noise.

I have tried to use covariance estimation that implemented [here](https://censi.science/software/csm/)
As I remember I've tried several things for the covariance matrix but every time I broke it more. Later, I got into different projects :D so I left it as it is.
